Tribler/Core/CreditMining/CreditMiningManager.py

```diff
@@ -0,0 +1,261 @@
+
+from Tribler.Core.CreditMining.CreditMiningSource import ChannelSource
+from Tribler.Core.CreditMining.CreditMiningPolicy import UploadPolicy, RandomPolicy
+from Tribler.Core.DownloadConfig import DefaultDownloadStartupConfig, DownloadStartupConfig
+from Tribler.Core.simpledefs import DLSTATUS_DOWNLOADING, DLSTATUS_STOPPED, DLSTATUS_SEEDING, DLSTATUS_STOPPED_ON_ERROR
+from Tribler.Core.TorrentDef import TorrentDefNoMetainfo
+from Tribler.dispersy.taskmanager import TaskManager
+
+
+class CreditMiningTorrent(object):
+    def __init__(self, infohash, name, download=None, state=None):
+        self.infohash = infohash
+        self.name = name
+        self.download = download
+        self.state = state
+        self.sources = set()
+        self.force_checked = False
+
+
+class CreditMiningSettings(object):
+    """
+    This class contains settings used by the credit mining manager
+    """
+    def __init__(self):
+        self.max_torrents_active = 8
+        self.max_torrents_listed = 100
+        # Note: be sure to set this interval to something that gives torrents a fair chance of
+        # discovering peers and uploading data
+        self.auto_manage_interval = 600
+        self.hops = 1
+        self.save_path = os.path.join(DefaultDownloadStartupConfig.getInstance().get_dest_dir(), 'credit_mining')
+
+
+class CreditMiningManager(TaskManager):
+    """
+    Class to manage all the credit mining activities
+    """
+
+    def __init__(self, session, settings=None, policies=None):
+        super(CreditMiningManager, self).__init__()
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self._logger.info('Starting CreditMiningManager')
+
+        self.session = session
+        self.settings = settings or CreditMiningSettings()
+
+        self.sources = {}
+        self.torrents = {}
+        self.policies = []
+
+        # Our default policy: 50% of torrents are selected by upload, 50% of torrents are selected randomly
+        self.policies = policies or [UploadPolicy(), RandomPolicy()]
+
+        if not os.path.exists(self.settings.save_path):
+            os.makedirs(self.settings.save_path)
+
+        self.select_lc = self.register_task('select_torrents', LoopingCall(self.select_torrents))
+        self.num_checkpoints = len(glob(os.path.join(self.session.get_downloads_pstate_dir(), '*.state')))
+
+        def add_sources(_):
+            for source in self.session.config.get_credit_mining_sources():
+                self.add_source(source)
+
+        self.session_ready = Deferred()
+        self.session_ready.addCallback(add_sources)
+
+    def shutdown(self, remove_downloads=False):
+        """
+        Shutting down credit mining manager. It also stops and remove all the sources.
+        """
+        self._logger.info('Shutting down CreditMiningManager')
+
+        deferreds = []
+        if remove_downloads:
+            sources = self.sources.keys()
+            for source in sources:
```

Comment:
Maybe compress it to `deferreds = [self.remove_source(source) for source in sources]`?

Reply:
Will do.